lgn-messages/src/types/v1/query/tasks.rs

```diff
@@ -161,20 +166,146 @@ pub struct SinglePathLeafInput
     pub embedded_proof: Vec<u8>,
 }
 
-#[derive(Clone, PartialEq, Dbg, Deserialize, Serialize)]
-pub struct RevelationInput
+#[derive(Clone, Dbg, Serialize, Deserialize)]
+pub struct HydratableMatchingRow
 {
-    pub placeholders: PlaceHolderLgn,
+    pub proof: Hydratable<db_keys::ProofKey>,
+    pub path: RowPath,
+    pub result: Vec<U256>,
+}
+impl HydratableMatchingRow
+{
+    pub fn into_matching_row(self) -> MatchingRow
+    {
+        MatchingRow::new(
+            self.proof
+                .clone_proof(),
+            self.path,
+            self.result,
+        )
+    }
+}
 
-    pub indexing_proof_location: db_keys::ProofKey,
+/// Either a `Dehydrated` variant containing a key to a stored proof, or a
+/// `Hydrated` containing the proof itself.
+#[derive(Clone, Serialize, Deserialize)]
+pub enum Hydratable<K: Clone + std::fmt::Debug>
+{
+    Dehydrated(K),
+    Hydrated(Arc<Vec<u8>>),
+}
 
-    pub query_proof_location: ProofKey,
+impl<T: Clone + std::fmt::Debug> std::fmt::Debug for Hydratable<T>
+{
+    fn fmt(
+        &self,
+        f: &mut std::fmt::Formatter<'_>,
+    ) -> std::fmt::Result
+    {
+        match self
+        {
+            Hydratable::Dehydrated(k) =>
+            {
+                write!(
+                    f,
+                    "dehydrated: {k:?}"
+                )
+            },
+            Hydratable::Hydrated(_) =>
+            {
+                write!(
+                    f,
+                    "hydrated"
+                )
+            },
+        }
+    }
+}
 
-    #[dbg(placeholder = "...")]
-    pub indexing_proof: Vec<u8>,
+impl<K: Clone + std::fmt::Debug> Hydratable<K>
+{
+    /// Wrap a proof key into a `Dehydrated` variant.
+    pub fn new(k: K) -> Self
+    {
+        Hydratable::Dehydrated(k)
+    }
 
-    #[dbg(placeholder = "...")]
-    pub query_proof: Vec<u8>,
+    /// Consume a `Hydrated` variant into its embedded proof; panic if it is
+    /// not hydrated.
+    pub fn proof(&self) -> Arc<Vec<u8>>
+    {
+        match self
+        {
+            Hydratable::Dehydrated(_) => unreachable!(),
+            Hydratable::Hydrated(proof) => proof.clone(),
+        }
+    }
+
+    /// Consume a `Hydrated` variant into its embedded proof; panic if it is
+    /// not hydrated.
+    pub fn clone_proof(&self) -> Vec<u8>
+    {
+        match self
+        {
+            Hydratable::Dehydrated(_) => unreachable!(),
+            Hydratable::Hydrated(proof) =>
+            {
+                proof
+                    .clone()
+                    .to_vec()
```

Comment:
Fine to clone the key but cloning the proof is ... annoying. Any chance we can get rid of ? 
A proof is like 150-200KB so when we deal with hundreds or thousands of proofs per minute, we're starting to add up a bit. 
It's definitely not at the range where it can be problematic but if we can avoid future problems now, even better...

Reply:
100% agreed; but the issue is that the prover interface requires to take ownership of the proof instead of accepting a slice of `u8`'s. I intend to convert it this way at some point, but I think these PRs are already chunky enough for now.